File: dp/level2/LC_Maximum_Subarray.java
package dp.level2;

public class LC_Maximum_Subarray {
	public int maxSubArray(int[] nums) {
		int dp[] = new int[nums.length];
		dp[0] = nums[0];
		int answer = dp[0];
		for (int i = 1; i < nums.length; i++) {
			dp[i] = Math.max(nums[i], dp[i-1] + nums[i]);
			answer = Math.max(answer, dp[i]);
		}

		return answer;
	}
}

File: dp/level2/LC_Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        N = len(nums)
        dp = [0] * N  # i인덱스를 마지막으로 포함하는 subarray의 최대 합
        dp[0] = nums[0]
        for i in range(1, N):
            if dp[i-1] + nums[i] < nums[i]:
                dp[i] = nums[i]
            else:
                dp[i] = dp[i-1] + nums[i]

        return max(dp)
